src/TuxState.js

```diff
@@ -0,0 +1,650 @@
+
+        }
+      }
+    }
+  };
+  recurseKeys(currentState, newState, newProps);
+  return newState;
+};
+
+
+var hasAnyOuterKeysMatching = function (currentState, newProps) {
+  for (var key in newProps) {
+    if (currentState[key]) {
+      return true;
+    }
+  }
+  invariant(!newProps, "No keys match any keys in the current state.");
+};
+
+var invariantNumberCheck = function (input) {
+  if (typeof(input) !== "number") {
+    invariant(!input, 'Cannot perform operation on "%s" because it is not of type number.', input);
+  };
+};
+
+var invariantArrayCheck = function (input, message) {
+  if (!Array.isArray(input)) {
+    invariant(!input, 'Cannot perform operation on "%s" because it is not an array.', input);
+  };
+};
+
+var invariantValueCheck = function (input) {
+  if (typeof(input) === "object") {
+    invariant(!input, 'Cannot perform operation on "%s" because it must be a single value.', input);
+  }
+}
+
+var invariantNumberOrStringCheck = function (input) {
+  if (typeof(input) !== "number" && typeof(input) !== "string") {
+    // console.log(typeof(input))
+    invariant(!input, 'Cannot perform operation on "%s" because it is not of type number or of type string.', input);
+  };
+};
+
+var invariantArgCheck = function (input) {
+  if (!input || typeof(input) !== 'object' || Array.isArray(input)) {
+    invariant(!input, 'This function requires an object as an argument.', input);
+  };
+}
+
+//mixin to that adds convenient methods for updating the state of a component
+var stateConvenienceMethods = {
+  state: {
+    'Pat': {
+      'cat': {
+        'name':'Mr. Bigglesworth',
+        'age':10
+      },
+      'dog':{
+        'age':10
+      },
+      'squirrel':true,
+      'pigeon':{
+        'fat':true
+      }
+    },
+    'Dmitri': {
+      'cat':{
+        'age': 12,
+        'name':'Spencer'
+      }
+    },
+    'Gunnari': {
+      'turtles':['Pat', 'Spencer']
+    }
+  },
+
+  setState: function(newState) {
+    this.state = assign({}, this.state, newState);
+  },
+
+  replaceState: function(newState) {
+    this.state = assign({}, newState);
+  },
+
+  // getInitialState: function () {
+  //   return {
+  //     'Pat':true,
+  //     'Dmitri':true,
+  //     'Gunnari':true,
+  //     'Spencer':true
+  //   };
+  // },
+
+  //addState FUNCTION: adds the values of the deepest keys in the passed in object to the corresponding deepest keys in the current state, or throws an error if keys don't match
+  //@param propsToAdd OBJECT: required object argument with deepest keys being numbers or strings to add to the current state
+  //@param callback FUNCTION: optional callback argument that will be executed once setState is completed and the component is re-rendered
+  addState: function (propsToAdd, callback) {
+    //throw error if propsToAdd is undefined or not an object
+    invariantArgCheck(propsToAdd);
+    //make a shallow copy of the current state
+    var currentState = assign({}, this.state);
```

Comment:
I don't think we need to invoke `assign` here because we are already making a shallow copy of state at the first line in `buildNewState`.